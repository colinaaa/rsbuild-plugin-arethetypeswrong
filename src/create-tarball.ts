import fs from "node:fs/promises";
import path from "node:path";

import { detect } from "package-manager-detector";
import { x } from "tinyexec";

interface Result extends AsyncDisposable {
  path: string;
}

export async function createTarball(
  root: string,
  pkg: { name: string; version: string },
): Promise<Result> {
  const tarballPath = path.join(root, `${pkg.name}-${pkg.version}.tgz`);

  const agent = await detect({ cwd: root, stopDir: root }) ?? { name: "npm" };

  await x(agent.name, ["pack"]);

  return {
    path: tarballPath,
    async [Symbol.asyncDispose]() {
      return fs.unlink(tarballPath);
    },
  };
}
