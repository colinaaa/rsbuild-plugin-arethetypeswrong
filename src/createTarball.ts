import fs from "node:fs/promises";
import path from "node:path";

import { detect } from "package-manager-detector";
import { x } from "tinyexec";

interface Result extends AsyncDisposable {
  path: string;
}

export async function createTarball(
  root: string,
  pkg: { name: string; version: string },
): Promise<Result> {
  const tarballPath = path.join(
    root,
    `${normalizeNpmPackageName(pkg.name)}-${pkg.version}.tgz`,
  );

  const { agent } = await detect({ cwd: root, stopDir: root }) ?? { agent: "npm" };

  const [command, args] = (function(): [string, string[]] {
    switch (agent) {
      case "yarn":
        // https://classic.yarnpkg.com/lang/en/docs/cli/pack/
        return [agent, ["pack", "--filename", tarballPath]];
      case "yarn@berry":
        // https://yarnpkg.com/cli/pack
        return ["yarn", ["pack", "--out", tarballPath]];
      case "npm":
      case "bun":
      case "deno":
        return ["npm", ["pack"]];
      case "pnpm":
      case "pnpm@6":
        return ["pnpm", ["pack"]];
    }
  })();

  await x(command, args, {
    nodeOptions: {
      cwd: root,
    },
  });

  return {
    path: tarballPath,
    async [Symbol.asyncDispose]() {
      return fs.unlink(tarballPath);
    },
  };
}

function normalizeNpmPackageName(name: string): string {
  return name.replaceAll("@", "").replaceAll("/", "-");
}
