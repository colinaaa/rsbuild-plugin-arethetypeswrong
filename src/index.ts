import fs from "node:fs";
import path from "node:path";

import type { RsbuildPlugin } from "@rsbuild/core";

import type { RenderOptions } from "./render/index.js";

export interface PluginAreTheTypesWrongOptions {
  /**
   * Whether to enable arethetypeswrong.
   * @default true
   */
  enable?: boolean;

  areTheTypesWrongOptions?: RenderOptions;
}

export const pluginAreTheTypesWrong = (
  options: PluginAreTheTypesWrongOptions = {},
): RsbuildPlugin => ({
  name: "plugin-arethetypeswrong",

  setup(api) {
    if (options.enable === false) {
      return;
    }

    const logger = api.logger;

    api.onAfterBuild({
      handler: async ({ isFirstCompile, isWatch }) => {
        // Only run on the first compile in watch mode, or on a single build
        if (!isFirstCompile) {
          return;
        }

        const { rootPath } = api.context;

        // Read package.json to get package name and version
        const packageJson = JSON.parse(
          fs.readFileSync(path.join(rootPath, "package.json"), "utf-8"),
        ) as { name: string; version: string };
        const packageName = packageJson.name;
        const packageVersion = packageJson.version;

        logger.start(`[arethetypeswrong] Checking ${packageName}@${packageVersion}...`);
        console.info();

        logger.debug(`[arethetypeswrong] Running npm pack from ${rootPath}`);
        const { createTarball } = await import("./createTarball.js");
        await using tarball = await createTarball(rootPath, packageJson);
        logger.debug(`[arethetypeswrong] npm pack success`);

        const { checkPackage, createPackageFromTarballData } = await import("@arethetypeswrong/core");
        const pkg = createPackageFromTarballData(fs.readFileSync(tarball.path));
        const result = await checkPackage(pkg);

        const { render } = await import("./render/index.js");

        const message = await render(result, options.areTheTypesWrongOptions ?? {});

        const { getExitCode } = await import("./getExitCode.js");

        const exitCode = getExitCode(result, options.areTheTypesWrongOptions);
        const hasErrors = exitCode !== 0;
        if (hasErrors) {
          logger.error(message);
          if (!isWatch) {
            throw new Error("arethetypeswrong failed!");
          }
          return;
        }
        logger.success(message);
      },
      order: "post",
    });
  },
});
