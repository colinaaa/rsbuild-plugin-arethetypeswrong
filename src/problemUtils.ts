import * as core from "@arethetypeswrong/core";
import type { ProblemKind } from "@arethetypeswrong/core";

export type ProblemFlag =
  | "no-resolution"
  | "untyped-resolution"
  | "false-cjs"
  | "false-esm"
  | "cjs-resolves-to-esm"
  | "fallback-condition"
  | "cjs-only-exports-default"
  | "named-exports"
  | "false-export-default"
  | "missing-export-equals"
  | "unexpected-module-syntax"
  | "internal-resolution-error";

export const problemFlags: Record<ProblemKind, ProblemFlag> = {
  NoResolution: "no-resolution",
  UntypedResolution: "untyped-resolution",
  FalseCJS: "false-cjs",
  FalseESM: "false-esm",
  CJSResolvesToESM: "cjs-resolves-to-esm",
  FallbackCondition: "fallback-condition",
  CJSOnlyExportsDefault: "cjs-only-exports-default",
  NamedExports: "named-exports",
  FalseExportDefault: "false-export-default",
  MissingExportEquals: "missing-export-equals",
  UnexpectedModuleSyntax: "unexpected-module-syntax",
  InternalResolutionError: "internal-resolution-error",
};

export const resolutionKinds: Record<core.ResolutionKind, string> = {
  node10: "node10",
  "node16-cjs": "node16 (from CJS)",
  "node16-esm": "node16 (from ESM)",
  bundler: "bundler",
};

export const moduleKinds = {
  1: "(CJS)",
  99: "(ESM)",
  "": "",
};
