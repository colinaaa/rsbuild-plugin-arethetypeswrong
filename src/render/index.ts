import type { CheckResult, ResolutionKind } from "@arethetypeswrong/core";

import type { ProblemFlag } from "../problemUtils.js";
import { renderTyped } from "./typed.js";
import { renderUntyped } from "./untyped.js";

export async function render(analysis: CheckResult, options: RenderOptions): Promise<string> {
  if (analysis.types) {
    return renderTyped(analysis, options);
  } else {
    return renderUntyped(analysis);
  }
}

export interface RenderOptions {
  ignoreRules?: ProblemFlag[];
  ignoreResolutions?: ResolutionKind[];
  summary?: boolean;
  emoji?: boolean;
}
