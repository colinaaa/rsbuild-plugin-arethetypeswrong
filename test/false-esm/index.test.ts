import { existsSync } from "node:fs";
import path from "node:path";
import { stripVTControlCharacters } from "node:util";

import { createRsbuild, logger } from "@rsbuild/core";
import { expect, test, vi } from "vitest";

import { pluginAreTheTypesWrong } from "../../src";

test("should throw when has false ESM", async () => {
  const rsbuild = await createRsbuild({
    cwd: import.meta.dirname,
    rsbuildConfig: {
      plugins: [pluginAreTheTypesWrong()],
    },
  });

  const error = vi.spyOn(logger, "error");

  await expect(rsbuild.build()).rejects.toThrowErrorMatchingInlineSnapshot(`[Error: arethetypeswrong failed!]`);

  expect(
    error.mock.calls.flatMap(call =>
      call
        .filter(message => typeof message === "string" && message.includes("[arethetypeswrong]"))
        .map(stripVTControlCharacters)
    ),
  ).toMatchSnapshot();

  expect(existsSync(path.join(import.meta.dirname, "test-false-esm-0.0.0.tgz"))).toBeFalsy();
});

// TODO: fix this
test.skip("should pass when all thrown resolution is disabled", async () => {
  const rsbuild = await createRsbuild({
    cwd: import.meta.dirname,
    rsbuildConfig: {
      plugins: [
        pluginAreTheTypesWrong({
          areTheTypesWrongOptions: {
            ignoreResolutions: [
              "node16-esm",
            ],
          },
        }),
      ],
    },
  });

  const success = vi.spyOn(logger, "success");

  const { close } = await rsbuild.build();

  expect(
    success.mock.calls.flatMap(call =>
      call
        .filter(message => typeof message === "string" && message.includes("[arethetypeswrong]"))
        .map(stripVTControlCharacters)
    ),
  ).toMatchSnapshot();

  expect(existsSync(path.join(import.meta.dirname, "test-false-esm-0.0.0.tgz"))).toBeFalsy();

  await close();
});

test("should be able to ignore rule false-esm", async () => {
  const rsbuild = await createRsbuild({
    cwd: import.meta.dirname,
    rsbuildConfig: {
      plugins: [
        pluginAreTheTypesWrong({
          areTheTypesWrongOptions: {
            ignoreRules: [
              "false-esm",
            ],
          },
        }),
      ],
    },
  });

  const success = vi.spyOn(logger, "success");

  const { close } = await rsbuild.build();

  expect(
    success.mock.calls.flatMap(call =>
      call
        .filter(message => typeof message === "string" && message.includes("[arethetypeswrong]"))
        .map(stripVTControlCharacters)
    ),
  ).toMatchSnapshot();

  expect(existsSync(path.join(import.meta.dirname, "test-false-esm-0.0.0.tgz"))).toBeFalsy();

  await close();
});

test("should not throw when enable: false", async () => {
  const rsbuild = await createRsbuild({
    cwd: import.meta.dirname,
    rsbuildConfig: {
      plugins: [
        pluginAreTheTypesWrong({
          enable: false,
        }),
      ],
    },
  });

  const success = vi.spyOn(logger, "success");

  const { close } = await rsbuild.build();

  expect(success).not.toBeCalled();

  expect(existsSync(path.join(import.meta.dirname, "test-false-esm-0.0.0.tgz"))).toBeFalsy();

  await close();
});
