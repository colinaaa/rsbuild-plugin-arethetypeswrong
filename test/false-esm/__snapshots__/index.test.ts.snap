// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to ignore rule false-esm 1`] = `
[
  "[arethetypeswrong] test-false-esm v0.0.0

 (ignoring rules: 'false-esm')

 No problems found 🌟
┌───────────────────┬──────────────────┐
│                   │ "test-false-esm" │
├───────────────────┼──────────────────┤
│ node10            │ 🟢               │
├───────────────────┼──────────────────┤
│ node16 (from CJS) │ 🟢 (ESM)         │
├───────────────────┼──────────────────┤
│ node16 (from ESM) │ 🟢 (ESM)         │
├───────────────────┼──────────────────┤
│ bundler           │ 🟢               │
└───────────────────┴──────────────────┘",
]
`;

exports[`should throw when has false ESM 2`] = `
[
  "[arethetypeswrong] test-false-esm v0.0.0

👺 Import resolved to an ESM type declaration file, but a CommonJS JavaScript file. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseESM.md


┌───────────────────┬────────────────────────┐
│                   │ "test-false-esm"       │
├───────────────────┼────────────────────────┤
│ node10            │ 🟢                     │
├───────────────────┼────────────────────────┤
│ node16 (from CJS) │ 👺 Masquerading as ESM │
├───────────────────┼────────────────────────┤
│ node16 (from ESM) │ 🟢 (ESM)               │
├───────────────────┼────────────────────────┤
│ bundler           │ 🟢                     │
└───────────────────┴────────────────────────┘",
]
`;
